enhanced-stack-trace-queries.md

```diff
@@ -0,0 +1,111 @@
+# Enhanced stack trace queries
+
+## Backround
+
+The SDKs have a built-in `__stack_trace` query that can be used display where the workflow is blocked at the time of the
+query (the last processed workflow task).
+
+The response to a `__stack_trace` query is a single string representing a list of stack individual stack traces
+separated by double newlines.
+
+Example from the TS SDK:
+
+```
+    at someInternalFunction (/path/to/workflows/my-workflow.ts:18:10)
+    at someWorkflowFunction (/path/to/workflows/my-workflow.ts:60:20)
+
+    at someInternalFunction2 (/path/to/workflows/my-workflow.ts:100:30)
+    at someWorkflowFunction2 (/path/to/workflows/my-workflow.ts:123:5)
+```
+
+While this is valuable for developers to be able to understand how their code correlates to event histories, it is quite
+limited since it requires looking up the actual code to get the full picture.
+
+## Proposal
+
+SDKs can optionally implement an `__enhanced_stack_trace` query (name TBD) where instead of returning the stack traces
+as a single string, they return a pre-defined structure, that structure will also include file sources.
+
+The web UI can use this response to show the workflow code and highlight locations where it is blocked.
+
+### Proposed structure
+
+```ts
+export interface SDKInfo {
+  name: string;
+  version: string;
+}
+
+/**
+ * Represents a slice of a file starting at lineOffset
+ */
+export interface FileSlice {
+  content: string;
+  lineOffset: number;
+}
+
+/**
+ * A pointer to a location in a file
+ */
+export interface FileLocation {
+  /**
+   * Path to source file (absolute or relative).
+   * When using a relative path, make sure all paths are relative to the same root.
+   */
+  filepath: string;
+  line: number;
+  column: number;
+  /**
+   * function/goroutine/thread/greenlet/whatever name, if it exists.
+   * (Not sure if this will be used but doesn't hurt to have some more information)
+   */
+  context?: string;
+}
+
+export interface StackTrace {
+  locations: FileLocation[];
+}
+
+/**
+ * Used as the result for the enhanced stack trace query
+ */
+export interface EnhancedStackTrace {
+  sdk: SDKInfo;
+  /**
+   * Mapping of file name to file contents
+   */
+  sources: Record<string, FileSlice[]>;
```

Comment:
Yes,

I'm now on the fence whether the SDK / UI should fallback to the old mechanism or omit some of the data here.
Maybe we should add a capabilities object to the response so the UI can adjust the view accordingly.

Reply:
:+1: It should be opt-in from the worker side I think. Shouldn't be able to issue a query to get source code by default.

Reply:
Hmmm.. would you opt-in to returning the code? Why? Is this because of the security concern with remote codecs?

Reply:
Yes, because of security concern in general (unrelated to remote codecs). IMO we should not have a built-in query that can return code. But I understand that can be annoying to ask developers to opt in to. I would welcome others' opinions here.

Reply:
I'm just curious why the code is different from data in that regard.
I'd add a worker option just in case, just not sure if the default should be on or off.

Reply:
Because it is clear to a user they are sending data. It is not clear there is a secret enabled by default backdoor to their source code. Remove the "enabled by default" from that statement and it becomes more reasonable.

Reply:
Clarified this is opt-in.